qp/utils.py

```diff
@@ -0,0 +1,46 @@
+import numpy as np
+import sys
+
+def safelog(arr, threshold=sys.float_info.epsilon):
+    """
+    Takes log of array with zeroes.
+
+    Parameters
+    ----------
+    arr: ndarray
+        Values to be logged
+    threshold: float
+        Small, positive value to replace zeros and negative numbers
+
+    Returns
+    -------
+    logged: ndarray
+        Logged values, with approximation in place of zeros and negative numbers
+    """
+    shape = np.shape(arr)
+    flat = arr.flatten()
+    logged = np.log(np.array([max(a,threshold) for a in flat])).reshape(shape)
+    return logged
+
+def calckl(qn, pn):
+    """
+    Calculates Kullback-Leibler Divergence
+
+    Parameters
+    ----------
+    qn: ndarray
```

Comment:
It would be nice to have it read like D=KL(P,Q) where P and Q are functions to be evaluated under the hood, but I'd have to specify the shared grid somewhere.  The calculate_kl_divergence function in utils.py now takes two functions and a grid, so it would read like D=KL(P,Q,grid), which is not quite as nice.  Maybe there's another way for me to think about the structure of this function, perhaps so it's not even in pdf.py but just in utils.py?